src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1325 @@
+		SET_NULLABLE_FIELD(task, pid, pid);
+
+		/* Update task status to indicate it is running */
+		UpdateBackgroundTask(task);
+		UpdateBackgroundJob(task->jobid);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		bool hadError = false;
+		StringInfoData message = { 0 };
+		initStringInfo(&message);
+
+		while (GetBackgroundWorkerPid(handle, &pid) != BGWH_STOPPED)
+		{
+			int latchFlags = WL_LATCH_SET | WL_TIMEOUT | WL_EXIT_ON_PM_DEATH;
+			int rc = WaitLatch(MyLatch, latchFlags, (long) 1000, PG_WAIT_EXTENSION);
+
+			if (rc & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, &message, &hadError);
+
+			shm_mq_detach(responseq);
+		}
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * There is a small race condition between the last loop and the backend being
+		 * terminated. We re-consume the output of the process here again in case we have
+		 * missed a message.
+		 */
+		{
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, &message, &hadError);
+
+			shm_mq_detach(responseq);
+		}
+
+		/*
+		 * Same as before, we need to lock pg_dist_background_task in a way where we can
+		 * check if there had been a concurrent cancel.
+		 */
+		LockRelationOid(DistBackgroundTaskRelationId(), ExclusiveLock);
+
+		oldContext = MemoryContextSwitchTo(perTaskContext);
+		task = GetBackgroundTaskByTaskId(task->taskid);
+		MemoryContextSwitchTo(oldContext);
+
+		if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING ||
+			task->status == BACKGROUND_TASK_STATUS_CANCELLED)
+		{
+			/*
+			 * A concurrent cancel has happened or the task has disappeared, we are not
+			 * retrying or changing state, we will only reflect the message onto the task
+			 * and for completeness we update the job aswell, this should be a no-op
+			 *
+			 * We still need to release the shared memory and xact before looping
+			 */
+
+			dsm_detach(seg);
+
+			task->status = BACKGROUND_TASK_STATUS_CANCELLED;
+			task->message = message.data;
+			UpdateBackgroundTask(task);
+			UpdateBackgroundJob(task->jobid);
+
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+
+			continue;
+		}
+		else if (hadError)
+		{
+			/*
+			 * When we had an error we need to decide if we want to retry (keep the
+			 * runnable state), or move to error state
+			 */
+			if (!task->retry_count)
+			{
+				SET_NULLABLE_FIELD(task, retry_count, 1);
```

Comment:
I feel like it's kind of silly that this is a nullable field, even though we always set it in the end. Can't it just be initialized at 0?

Reply:
Currently we don't always set it. If the task completes normally on the first run we keep this value `NULL` which keeps the catalog table relatively empty. This makes it quite easy to spot check if there have been failures when running things in the background.

I can make this field default to `0` if you feel the empty nature of the field doesn't add as much value as I assumed. The code would become somewhat easier in this area.